### PR TITLE
TMDM-12544 Security issue: XXE injection

### PR DESCRIPTION
**Pull request by hw-zhu**
**What is the current behavior?** (You should also link to an open issue here)

there is security issue to parse xml using third xml library

**What is the new behavior?**

resolve above issue.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/XMLUtils.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private static final Logger LOGGER = Logger.getLogger(XMLUtils.class);
+
+    public static boolean isExistExtEntity(InputStream stream) {
+        boolean results = false;
+        try {
+            String result = new BufferedReader(new InputStreamReader(stream))
+                    .lines().collect(Collectors.joining(System.lineSeparator()));
+            results = isExistExtEntity(result);
+        } catch (Exception e) {
+            LOGGER.error("An unexpected exception occurred caused by " + e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/XMLUtils.java]: Change to `LOGGER.error("An unexpected exception occurred.", e);`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/exception/XmlBeanDefinitionException.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * %InstallDIR%\features\org.talend.rcp.branding.%PRODUCTNAME%\%PRODUCTNAME%license.txt
+ * 
+ * You should have received a copy of the agreement along with this program; if not, write to Talend SA 9 rue Pages
+ * 92150 Suresnes, France
+ */
+package org.talend.mdm.commmon.util.exception;
+
+public class XmlBeanDefinitionException extends RuntimeException {
+
+    private static final long serialVersionUID = 1L;
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/exception/XmlBeanDefinitionException.java]: Use machine generated value for `serialVersionUID`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static final String FEATURE_EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+
+    public static final String FEATURE_EXTERNAL_PARAM_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+
+    public static final String FEATURE_DEFER_NODE_EXP = "http://apache.org/xml/features/dom/defer-node-expansion";
+
+    public static final String PROPERTY_IS_SUPPORT_EXT_ENTITY = "javax.xml.stream.isSupportingExternalEntities";
+
+    public static final String PROPERTY_SCHEMAL_LANGUAGE = "http://java.sun.com/xml/jaxp/properties/schemaLanguage";
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java]: 'PROPERTY_SCHEMAL_LANGUAGE' should be 'PROPERTY_SCHEMA_LANGUAGE', right?

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import org.talend.mdm.commmon.util.exception.XmlBeanDefinitionException;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.DefaultHandler;
+import org.xml.sax.helpers.XMLReaderFactory;
+
+public class MDMXMLUtils {
+
+    private static final Logger LOGGER = Logger.getLogger(MDMXMLUtils.class);
+
+    private static final DocumentBuilderFactory DOM_BUILDER_FACTORY;
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `DOC_BUILDER_FACTORY;` would be better.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    private static final DocumentBuilderFactory DOM_BUILDER_FACTORY_WITH_NAMESPACE;
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `DOC_BUILDER_FACTORY_WITH_NAMESPACE;` would be better.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        XML_INPUT_FACTORY.setProperty(XMLInputFactory.IS_NAMESPACE_AWARE, Boolean.TRUE);
+        XML_INPUT_FACTORY.setProperty(XMLInputFactory.IS_COALESCING, Boolean.TRUE);
+        XML_INPUT_FACTORY.setProperty(MDMXMLConstants.PROPERTY_IS_SUPPORT_EXT_ENTITY, false);
+    }
+
+    public static XMLStreamReader createXMLStreamReader(InputStream inputStream) {
+        try {
+            return XML_INPUT_FACTORY.createXMLStreamReader(inputStream);
+        } catch (Exception e) {
+            throw new XmlBeanDefinitionException("Error occurred creating a new XMLStreamReader from a java.io.InputStream.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while creating a new XMLStreamReader from InputStream.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+
+    public static XMLEventReader createXMLEventReader(InputStream inputStream) {
+        try {
+            return XML_INPUT_FACTORY.createXMLEventReader(inputStream);
+        } catch (Exception e) {
+            throw new XmlBeanDefinitionException("Error occurred creating a new XMLEventReader from a InputStream.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while creating a new XMLEventReader from InputStream.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+
+    public static XMLEventReader createXMLEventReader(StringReader source) {
+        try {
+            return XML_INPUT_FACTORY.createXMLEventReader(source);
+        } catch (Exception e) {
+            throw new XmlBeanDefinitionException("Error occurred creating a new XMLEventReader from a StringReader.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while creating a new XMLEventReader from StringReader.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+
+    public static SAXParser getSAXParser() {
+        try {
+            return SAX_PARSER_FACTORY.newSAXParser();
+        } catch (Exception e) {
+            throw new XmlBeanDefinitionException("Error occurred using SAXParserFactory to create a SAXParser.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while creating a SAXParserFactory from SAXParser.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    private static DocumentBuilderFactory getDocumentBuilderFactoryWithNamespace() {
+        return DOM_BUILDER_FACTORY_WITH_NAMESPACE;
+    }
+
+    public static Optional<DocumentBuilder> getDocumentBuilder() {
+        DocumentBuilder builder = null;
+        try {
+            builder = getDocumentBuilderFactory().newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new XmlBeanDefinitionException("Error occurred using DocumentBuilderFactory to create a DocumentBuilder.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while using DocumentBuilderFactory to create a DocumentBuilder.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        return Optional.ofNullable(builder);
+    }
+
+    public static Optional<DocumentBuilder> getDocumentBuilderWithNamespace() {
+        DocumentBuilder builder = null;
+        try {
+            builder = getDocumentBuilderFactoryWithNamespace().newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new XmlBeanDefinitionException("Error occurred using DocumentBuilderFactory to create a DocumentBuilder.", e);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: `Error occurred while using DocumentBuilderFactory to create a DocumentBuilder.`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static final String FEATURE_LOAD_EXTERNAL = "http://apache.org/xml/features/nonvalidating/load-external-dtd";
+
+    public static final String FEATURE_EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+
+    public static final String FEATURE_EXTERNAL_PARAM_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+
+    public static final String FEATURE_DEFER_NODE_EXP = "http://apache.org/xml/features/dom/defer-node-expansion";
+
+    public static final String PROPERTY_IS_SUPPORT_EXT_ENTITY = "javax.xml.stream.isSupportingExternalEntities";
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java]: Would suggest to change variable name to 'PROPERTY_IS_SUPPORT_EXTERNAL_ENTITIES'.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static final String FEATURE_DISALLOW_DOCTYPE = "http://apache.org/xml/features/disallow-doctype-decl";
+
+    public static final String FEATURE_LOAD_EXTERNAL = "http://apache.org/xml/features/nonvalidating/load-external-dtd";
+
+    public static final String FEATURE_EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+
+    public static final String FEATURE_EXTERNAL_PARAM_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+
+    public static final String FEATURE_DEFER_NODE_EXP = "http://apache.org/xml/features/dom/defer-node-expansion";
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLConstants.java]: Would suggest to change variable name to 'FEATURE_DEFER_NODE_EXPANSION'.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+        try {
+            XML_READER = XMLReaderFactory.createXMLReader();
+            XML_READER.setFeature(MDMXMLConstants.FEATURE_DISALLOW_DOCTYPE, true);
+            XML_READER.setFeature(MDMXMLConstants.FEATURE_LOAD_EXTERNAL, false);
+            XML_READER.setFeature(MDMXMLConstants.FEATURE_EXTERNAL_GENERAL_ENTITIES, false);
+            XML_READER.setFeature(MDMXMLConstants.FEATURE_EXTERNAL_PARAM_ENTITIES, false);
+            XML_READER.setContentHandler(new DefaultHandler());
+        } catch (Exception e) {
+            throw new XmlBeanDefinitionException("Error occurred to initialize xmlReader", e);
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: Error occurred to initialize XMLReader.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+    public static final String FEATURE_EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+
+    public static final String FEATURE_EXTERNAL_PARAM_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+
+    public static final String FEATURE_DEFER_NODE_EXPANSION = "http://apache.org/xml/features/dom/defer-node-expansion";
+
+    public static final String PROPERTY_IS_SUPPORT_EXTERNAL_ENTITIES = "javax.xml.stream.isSupportingExternalEntities";
+
+    public static final String PROPERTY_SCHEMAL_LANGUAGE = "http://java.sun.com/xml/jaxp/properties/schemaLanguage";
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: Would it be PROPERTY_SCHEMA_LANGUAGE?

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+    public static final String PROPERTY_XML_SCHEMA = "http://www.w3.org/2001/XMLSchema";
+
+    public static final String PROPERTY_SCHEMAL_SOURCE = "http://java.sun.com/xml/jaxp/properties/schemaSource";
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/util/core/MDMXMLUtils.java]: Would it be PROPERTY_SCHEMA_SOURCE?

**Pull request merged**